code/datums/holocall.dm

```diff
@@ -0,0 +1,158 @@
+	var/list/dialed_holopads	//all things called, will be cleared out to just connected_holopad once answered
+
+	var/mob/camera/aiEye/remote/holo/eye	//user's eye, once connected
+	var/obj/effect/overlay/holo_pad_hologram/hologram	//user's hologram, once connected
+
+	var/call_start_time
+
+//creates a holocall made by `caller` from `calling_pad` to `callees`
+/datum/holocall/New(mob/living/caller, obj/machinery/hologram/holopad/calling_pad, list/callees)
+	call_start_time = world.time
+	user = caller
+	calling_pad.outgoing_call = src
+	calling_holopad = calling_pad
+	dialed_holopads = list()
+
+	for(var/I in callees)
+		var/obj/machinery/hologram/holopad/H = I
+		if(!qdeleted(H))// && !(H.stat & NOPOWER))
+			dialed_holopads += H
+			LAZYADD(H.holo_calls, src)
+
+	if(!dialed_holopads.len)
+		calling_holopad.atom_say("Connection failure.")
+		qdel(src)
+		return
+
+//cleans up ALL references :)
+/datum/holocall/Destroy()
+	if(!qdeleted(user))
+		user.reset_perspective()
+		if(user.client)
+			for(var/datum/camerachunk/chunk in eye.visibleCameraChunks)
+				chunk.remove(eye)
+		user.remote_control = null
+		user = null
+	QDEL_NULL(eye)
+
+	if(hologram)
+		hologram.HC = null
+		hologram = null
+
+	for(var/I in dialed_holopads)
+		var/obj/machinery/hologram/holopad/H = I
+		LAZYREMOVE(H.holo_calls, src)
+	dialed_holopads.Cut()
+
+	if(calling_holopad)
+		calling_holopad.outgoing_call = null
+		calling_holopad.SetLightsAndPower()
+		calling_holopad = null
+	if(connected_holopad)
+		connected_holopad.SetLightsAndPower()
+		connected_holopad = null
+
+
+	return ..()
+
+//Gracefully disconnects a holopad `H` from a call. Pads not in the call are ignored. Notifies participants of the disconnection
+/datum/holocall/proc/Disconnect(obj/machinery/hologram/holopad/H)
+	if(H == connected_holopad)
+		calling_holopad.atom_say("[usr] disconnected.")
+	else if(H == calling_holopad && connected_holopad)
+		connected_holopad.atom_say("[usr] disconnected.")
+
+	user.unset_machine(H)
+
+	ConnectionFailure(H, TRUE)
+
+//Forcefully disconnects a holopad `H` from a call. Pads not in the call are ignored.
+/datum/holocall/proc/ConnectionFailure(obj/machinery/hologram/holopad/H, graceful = FALSE)
+	if(H == connected_holopad || H == calling_holopad)
+		if(!graceful && H != calling_holopad)
+			calling_holopad.atom_say("Connection failure.")
+		qdel(src)
+		return
+
+	LAZYREMOVE(H.holo_calls, src)
+	dialed_holopads -= H
+	if(!dialed_holopads.len)
+		if(graceful)
+			calling_holopad.atom_say("Call rejected.")
+		qdel(src)
+
+//Answers a call made to a holopad `H` which cannot be the calling holopad. Pads not in the call are ignored
+/datum/holocall/proc/Answer(obj/machinery/hologram/holopad/H)
+	if(H == calling_holopad)
+		CRASH("How cute, a holopad tried to answer itself.")
+
+	if(!(H in dialed_holopads))
+		return
+
+	if(connected_holopad)
+		CRASH("Multi-connection holocall")
```

Comment:
This is one of the more irritating parts of Bay's code. Why does it crash this entire execution tree if it manages to get here? It ought to just return and resolve any set variables gracefully.